### DISCUSSION TITLE
Microsoft.Office.Interop.Excel behaviour change after migrating from packages.config to PackageReference

### DISCUSSION THREAD

**Discussion by JeremiCurie** [NuGet/Home#11664] [Q&A]
Hello,
After migrating from packages.config to PackageReference in a solution, I have an error when using the object "Value" of a cell in an Excel file :

example, when using this :
String test = (String)worksheet.Cells[1, 2].Value;

I get the following error :
error code : CS1061
error description : 'object' does not contain a definition for 'Value' and no accessible extension method 'Value' accepting a first argument of type 'object' could be found (are you missing a using directive or an assembly reference?)

I know I could rollback but I'd like to understand what's wrong after migration ?

thanks for your help.
Jérémi.

**Comment by KalleOlaviNiemitalo** [comment]
EmbedInteropTypes metadata is not automatically set for assembly references that come from PackageReference. That then causes worksheet.Cells[1, 2] to return object rather than dynamic. According to <https://github.com/NuGet/Home/issues/2365>, the package should include a targets file that sets the metadata, if the author of the package thinks the types should be embedded. If the package lacks that, you can add a similar target in your project or in Directory.Build.targets. <https://github.com/NuGet/Samples/blob/d185393502edf11f6ceecdae319a7d7c19b30eca/NuGet.Samples.Interop/NuGet.Samples.Interop.targets>

```XML
  <Target Name="EmbedExcelInterop" AfterTargets="ResolveReferences" BeforeTargets="FindReferenceAssembliesForReferences">
    <ItemGroup>
      <ReferencePath Condition="%(ReferencePath.NuGetPackageId) == 'Microsoft.Office.Interop.Excel'">
        <EmbedInteropTypes>true</EmbedInteropTypes>
      </ReferencePath>
    </ItemGroup>
  </Target>
```

**Reply by KalleOlaviNiemitalo** [reply]
Documentation for how to create COM interop NuGet packages was added to docs.microsoft.com in <https://github.com/NuGet/docs.microsoft.com-nuget/pull/506> and is currently at <https://docs.microsoft.com/nuget/create-packages/author-packages-with-com-interop-assemblies>.
